pkg/controller/quota/quotaforestmanager/qm_lib_backend_with_quotasubt_mgr.go

```diff
@@ -569,9 +570,41 @@ func (qm *QuotaManager) Fits(aw *arbv1.AppWrapper, awResDemands *clusterstateapi
 		return doesFit, preemptIds, strings.TrimSpace(allocResponse.GetMessage())
 	}
 	preemptIds = qm.getAppWrappers(allocResponse.GetPreemptedIds())
+
+	// Update cluster resources in the even that preemption happens
+	if clusterResources != nil {
+		updatedResources := clusterResources
+
+		for _, appWrapper := range preemptIds {
+			updatedResources.Add(qm.getAggregatedResources(appWrapper))
+		}
+
+		// Check if job fits with the update resources after preempted AppWrappers are removed
+		if clusterResources != nil && !awResDemands.LessEqual(updatedResources) {
+			qm.quotaManagerBackend.UndoAllocateForest(QuotaManagerForestName, consumerID)
+			qm.removeConsumer(consumerID)
+			return false, preemptIds, fmt.Sprintf("[Fits] AppWrapper '%s/%s' does not fit in the cluster, even after borrowed quota is freed", aw.Namespace, aw.Name)
+		}
+	}
+
 	return doesFit, preemptIds, strings.TrimSpace(allocResponse.GetMessage())
 }
 
+func (qm *QuotaManager) getAggregatedResources(appWrapper *arbv1.AppWrapper) *clusterstateapi.Resource {
+	allocated := clusterstateapi.EmptyResource()
+
+	for _, genericItem := range appWrapper.Spec.AggrResources.GenericItems {
```

Comment:
this is MCAD specific logic, I think this should be out of quota management library